### ISSUE TITLE
[performance] Symfony Kernel::boot() in dev mode on Alpine

### ISSUE THREAD

**Issue by bastnic** [symfony/symfony#35009]
This is a follow-up of [this discussion](https://github.com/symfony/recipes/pull/705#discussion_r358711700).

Alpine is the default docker linux base distribution. It's widely used. It's based on musl libc, and so, it's quite different of traditional linux distribution as Debian. One the difference is that glob does not handle `GLOB_BRACE`.

I knew about it from [an old issue on composer](https://github.com/composer/composer/pull/7444) that we already experienced. 

On a big project, we had quite castrophous boot time, and I finally checked on it. 

The first thing is that `Iterator/SortableIterator.php/44-46` cost an enormous amout of time and is called 22k time.

Searching why, It seems that we never enter is the optimized path here https://github.com/symfony/symfony/blob/8ffdef2958d4190994d7540013925b9580e03493/src/Symfony/Component/Config/Resource/GlobResource.php#L102.

I had that in `Kernel::configureContainer`:

```php
const CONFIG_EXTS = '.{php,xml,yaml,yml}';
// ... 
$loader->load($confDir . '/{packages}/*' . self::CONFIG_EXTS, 'glob'); 
$loader->load($confDir . '/{packages}/' . $this->environment . '/**/*' . self::CONFIG_EXTS, 'glob');
$loader->load($confDir . '/{services}' . self::CONFIG_EXTS, 'glob');
```

and in `services.yaml`:

```yaml
Lib\Core\:
    resource: '../src/*'
    exclude: '../src/{Entity,Migrations,Translation,Tests,Kernel.php,PHPStan,Fixtures,Traits}'
```

As I don't have `GLOB_BRACE` and every pattern given contains a brace, it always fallback on the `symfony/finder` implementation which is wayyy slower.

By removing the `/**/*`, and only searching into `yaml` extension, I expect to be in the fast path. But the brackets on `{packages}` and `{services}` prevents it. By removing them it goes into the fast path. And gives a clear boost.

```php
$loader->load($confDir . '/packages/*' . self::CONFIG_EXTS, 'glob');

if (is_dir($confDir . '/packages/' . $this->environment)) {
	$loader->load($confDir . '/packages/' . $this->environment . '/*' . self::CONFIG_EXTS, 'glob');
}
$loader->load($confDir . '/services.yaml');
```

I didn't see any drwabacks (and I have a very extensive tests suites).

Rest the case with multiple option separated by comma in brackets, that stills cost a lot.

I'm absolutly thinks this is a serious issue, but I'm not sure of what todo with it, hance the absence of PR.

- brackets are they really necessary for packages and services? (**edit** answer: it ONLY ensure that the file / dir is optional. On an app that is older than one day, I'm sure of the structure that I use, and it should be totally recommended in perf reco to update that Kernel stuff to be exact and don't let it guess.)
- should we purpose a better implementation on that hot path than the `symfony/finder` one. @nicolas-grekas purpose to "we could have some logic that splits a single glob-with-braces into several of them" https://github.com/symfony/recipes/pull/705#discussion_r358748699

**Comment by nicolas-grekas** [comment]
About my suggestion to splits single glob-with-braces, here is the description from the gnu doc:
https://www.gnu.org/software/libc/manual/html_node/More-Flags-for-Globbing.html

**Comment by bastnic** [comment]
I tries with [this implementation](https://github.com/jeedom/core/pull/957/files), and I get a stunning optimization of -202ms... (-64%)

https://blackfire.io/profiles/compare/3ff797c4-a931-4f18-8b4e-a57982d3d2cd/graph


![image](https://user-images.githubusercontent.com/84887/71009258-3614f280-20ea-11ea-97f3-a50265e6be36.png)

**Comment by bastnic** [comment]
Thanks a lot @nicolas-grekas!

**Comment by Sh1d0w** [comment]
@nicolas-grekas I think your fix broke the discovery of package configurations. If I apply your changes from (https://github.com/symfony/symfony/commit/d7a06790119b04e4f3e96f36dbf68e70ec5964ae) I get the following error:

`The child node "defaults" at path "overblog_dataloader" must be configured.`

If I remove the changes everything works fine.

**Comment by nicolas-grekas** [comment]
Can you please provide a reproducer?

**Comment by Sh1d0w** [comment]
@nicolas-grekas  Sorry, can't provide atm as the project we are working on is private and I am currently on a vacation.

The package that errors is https://packagist.org/packages/overblog/dataloader-bundle

We have the full configuration file set up with the defaults key as well. 

We are running our app with Docker: `FROM alpine:3.10.2`

The error `The child node "defaults" at path "overblog_dataloader" must be configured.` is thrown only after I apply the changes from your commit, if I remove them everything works fine.

I will try to put something as POC these days, but in the meantime I think thats enough info to reproduce.

**Comment by nicolas-grekas** [comment]
@Sh1d0w I am able to reproduce the error, but it looks unrelated to this change. I still have the error with v4.4.1.

> If I remove the changes everything works fine.

is this the only change you're doing or are there any others, like updating the bundle also?

**Comment by Sh1d0w** [comment]
@nicolas-grekas I've upgraded from Symfony 4.3 to 4.4 .

The bundle that errors is using the same version v0.4.2 and I did not update its version.

After the update to Symfony 4.4 I tested the app and it worked fine. Then tried to apply the change from your commit as I was keen to try it out before it was published, and I got this error. I've removed the changes to the GlobResource everything works fine.

Just to be clear if I revert to Symfony 4.3 where everything is working ok as well, and apply the same changes to GlobResource I get the same error.

**Comment by nicolas-grekas** [comment]
Can you share the output or print_r($paths) just before this line, which exists in both versions?
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Config/Resource/GlobResource.php#L118

**Comment by Sh1d0w** [comment]
@nicolas-grekas Actually your fix is working fine, the problem was on my end! Really sorry for the spam and thanks for the great fix :)

**Comment by bastnic** [comment]
@Sh1d0w good news. Can you share the perf improvement on your side? It's very impressive on my project but we miss other feedback.

**Comment by Sh1d0w** [comment]
Sure, it resulted approximately 50% reduction in load times, which is more than great (previously ~700ms, after the fix ~300 ms)

**Comment by bastnic** [comment]
@Sh1d0w wooot!

**Comment by Sh1d0w** [comment]
@nicolas-grekas Will this fix be back ported and released for older versions of Symfony as a patch, or will be available only for 4.4 and above?

**Comment by nicolas-grekas** [comment]
@Sh1d0w that's going to be for 4.4 and up only. One more reason to upgrade asap :)
